### DISCUSSION TITLE
RFC: HTML Fallback Handling

### DISCUSSION THREAD

**Discussion by dillonkearns** [dillonkearns/elm-markdown#99] [Ideas]
# HTML Fallback Handling

## Background

[GFM has a feature to disallow specific raw HTML](https://github.github.com/gfm/#disallowed-raw-html-extension-).

Should there be different ways to handle HTML tags with no HTML handler?

Some motivating cases:

- Ignore HTML tags rather than giving errors
    - Print as raw text?
    - Don't render at all?
    - Custom fallback rendering?
- Allowlist and/or blocklist helper function to render tags literally that match a specific value? Or maybe just a function that takes a list of tag names, list of attr names, and turns it into HTML, or tags list of tags and attributes and returns `type RawHtmlAllowed = Allowed | Disallowed`?
- Ideally, the parsing step should succeed independent of HTML handlers - then it would be a renderer concern as to whether to have an error with certain parsed HTML tags or not (just passing the parsed AST to the renderer).

**Comment by wolfadex** [comment]
At work the markdown parser we use (a JS based one) defaults to escaping all html. This was fine until I proposed switching from using handlebars syntax for custom rendering to using a combination of vanilla html and custom elements. Turns out that the parser doesn't support html by default (which is nice), but it also doesn't support custom elements at all!

Based on this experience I think an allow list would be the best approach for supporting html. Let the user define what they want.

**Comment by houseofmercy-github** [comment]
I really like the way Jim Carlson handles something like this in his L1 markup demonstration https://www.youtube.com/watch?v=gqeqtZeDp18&t=120s - when his parser encounters unsupported syntax it simply adds an error object and keeps parsing.  His renderer colors the error red along with the other rendered output so it's clear where the problem is.  Am currently trying some experiments to see if I can find a way to simulate this with a catchall hander.

**Reply by houseofmercy-github** [reply]
I managed to get something approximating my idea to work here
https://github.com/houseofmercy-github/elm-markdown/commit/d2b033e7ab026506004bfecf8e46ba00c8e13c1d#commitcomment-70848714

Unfortunately I wasn't able to figure out how to do this without changing **Markdown/Html.elm**
so I just did it that way.  I'm still learning the rules of Elm but It seems code outside of the Markdown 
module can't import Markdown.HtmlRenderer :-(

**Reply by dillonkearns** [reply]
Hello! This is very interesting, thanks for sharing that code.

For production code, we would probably do fallback behavior that's not a red error message, but I think your code gives a nice proof of concept of providing a fallback! And I think it would be a good idea to explore possible API designs for letting the user choose how to present errors - possibly letting the user use a custom renderer for nodes with errors.
